README.org
#+html: <p align="center"><img src="logo/islisp-black-tr.gif" /></p>
* islisp-mode

** Features
   + ISLisp major-mode with custom map keybindings
   + Inferior-mode for REPL integration 
   + Syntax highlighting
   + Interactive REPL commands (eg: ~islisp-eval-defun~)
   + Interactive ISLispHyperDraft documentation search
   + Implementations support (Easy-ISLisp)
   + [[https://www.emacswiki.org/emacs/ImenuMode][Imenu]] support
   + Advance features (autocompletions, tags support and symbol navigation)

** Install 
*** Source
To install, put this repository somewhere in your Emacs load path, and add a require to your main Emacs elisp file:
#+BEGIN_SRC emacs-lisp
(require 'islisp-mode)
#+END_SRC

To add a custom location to the load path:
#+BEGIN_SRC emacs-lisp
   (add-to-list 'load-path "/path/to/islisp-mode/")
#+END_SRC

** Usage
   + Example video, [[https://vimeo.com/614514131][here]]

   The major mode will detect .lsp extension files as ISLisp files and start the syntax highlighting.

   It is recommended to use ~inferior-islisp-mode~ alongside ~islisp-mode~ for better integration.

   A normal workflow can be:
   + Open or create a .lsp file
   + Launch the repl with  =M-x= ~islisp-repl~
     This will initialize the inferior mode, and load a ISLisp REPL.

     The defaul implementation executable is [[https://github.com/sasagawa888/eisl][Easy-ISLisp]], it can be change in ~inferior-islisp-command-line~.

   + Now, back to the .lsp buffer, you can start to send code to the REPL, (eg: with ~islisp-eval-defun~)

*** Configuration example
    #+BEGIN_SRC emacs-lisp
      (add-to-list 'load-path "/path/to/islisp-mode/")
      (require 'islisp-mode)

      (setf
      ;;; For local HyperDraft navigation
       islisp-hyperdraft-root "file:/home/fermin/Programming/ISLispHyperDraft/islisp-v23.html"
      ;;; Activate the advance features
       islisp-advance-features t)
    #+END_SRC


** Main commands

  | Key Combination | Function Name               | Explanation                                                                  |
  |-----------------+-----------------------------+------------------------------------------------------------------------------|
  | C-M-x           | islisp-eval-defun           | Send the current defun to the inferior ISLisp process.                       |
  | C-c C-k         | islisp-compile-file         | Compile a ISLisp file in the inferior ISLisp process. (Easy-ISLisp specific) |
  | C-c C-l         | islisp-load-file            | Load a ISLisp file into the inferior ISLisp process.  (Easy-ISLisp specific) |
  | C-c C-r         | islisp-eval-region          | Send the current region to the inferior ISLisp process.                      |
  | C-c M-q         | indent-region               | Indent the current region                                                    |
  | C-c C-d         | islisp-lookup-documentation | Search for documentation in ISLispHyperDraft                                 |
  | C-c TAB         | islisp-repl                 | Start a ISLisp REPL or switch to it.                                         |
  | C-x C-e         | islisp-eval-last-sexp       | Send the previous sexp to the inferior ISLisp process.                       |

   


** Implementations
   Currently, the only supported inplementation is [[https://github.com/sasagawa888/eisl][Easy-ISLisp]], more will be supported in the future.

*** Easy-ISLisp
     [[https://github.com/sasagawa888/eisl][Easy-ISLisp(EISL)]] is an interpreter and compiler compatible with the ISLisp standard. EISL was written by Kenichi Sasagawa. 

**** Features
     + File compilation
     + Buffer formatting
     + Macroexpand
     + Specific keyword syntax highlighting

**** Specific commands

  | Key Combination | Function Name                  | Explanation                                                        |
  |-----------------+--------------------------------+--------------------------------------------------------------------|
  | C-c M-m         | easy-islisp-macroexpand-sexp   | Macroexpand the current sexp.                                      |
  | C-c  M-r        | easy-islisp-macroexpand-region | Macroexpand the current region                                     |
  | C-c C-f         | easy-islisp-format-region      | Format the selected region using the easy-islisp library formatter |
  | C-c C-b         | easy-islisp-format-buffer      | Format the current buffer using the easy-islisp libray formatter   |
  
** Advance features
   This mode also incluse some "advance" features that enhance the basic major mode functionaly.

   IMPORTANT: This features are NOT enable by default, to enable it, you should set the variable ~islisp-advance-features~ to ~t~.

*** Tags integration

  | Key Combination | Function Name                | Explanation                                            |
  |-----------------+------------------------------+--------------------------------------------------------|
  | C-c C-w         | islisp-tags-symbols-navigate | Find the workspace TAG definition                      |
  | C-c TAB         | islisp-tags-autocomplete     | Show auto-completion at point using current tags table |
  | C-c C-e         | islisp-tags-generate         | Generate current project tags, created with `etags'.   |



** License
#+begin_example
  General Public License Version 3 (GPLv3)
  Copyright (c) Fermin MF - https://sasanidas.gitlab.io/f-site/
  [[file:LICENSE]]
#+end_example


CHANGELOG.org
* Changelog
All notable changes to this project will be documented in this file.

The format is based on [[https://keepachangelog.com/en/1.0.0/][Keep a Changelog]].






# ** [0.7.4] - 2020-11-06
# *** Added
# + Decouple =maxima-get-info-on-subject= from the global auxiliary process
# + The test section of the Makefile has been rewrite, make it suitable for CI/CD
# + The .gitignore file
# + Added CHANGELOG.org
# + Added Gitlab CI/CD integration

# *** Changed
# + Now the test are self contained, can be executed without leaving any background process.
# + Some tests are not active, this is due to a CI/CD time problem

# *** Fixed
# + Fix maxima-get-completions function in the company integration
# + Compatibility with Emacs 26.3

# *** Removed
# + hippie-expand functions 

# [0.7.4]: https://gitlab.com/sasanidas/maxima/-/tags/0.7.4


test/make-test.el
(let* ((project-tests-file "test-islisp-mode.el")
       (current-directory (file-name-directory load-file-name))
       (project-test-path (expand-file-name "." current-directory))
       (project-root-path (expand-file-name ".." current-directory)))

  ;; add the package being tested to 'load-path so it can be 'require-d
  (add-to-list 'load-path project-root-path)
  (add-to-list 'load-path project-test-path)

  ;; load the file with tests
  (load (expand-file-name project-tests-file project-test-path) nil t))


test/make-install.el
;; List of the all the dependencies, including the dev dependencies
(defconst dev-packages '(ert package-lint))

;; Initialize package.el
(setq package-user-dir
      (expand-file-name (format ".elpa/%s/elpa" emacs-version)))

(message "Installing in %s ...\n" package-user-dir)
(package-initialize)

(setq package-archives
      '(("melpa" . "https://melpa.org/packages/")
        ("gnu" . "https://elpa.gnu.org/packages/")))

(package-refresh-contents)

;; Install dependencies
(seq-map (lambda (package)
	   (unless (package-installed-p package)
	     (ignore-errors
	       (package-install package))))
	 dev-packages)

;; Upgrade dependencies
(save-window-excursion
  (package-list-packages t)
  (condition-case nil
      (progn
        (package-menu-mark-upgrades)
        (package-menu-execute t))
    (error
     (message "All packages up to date"))))


test/test-islisp-mode.el
;;; test-islisp-mode.el --- islisp-mode test file.               -*- lexical-binding: t; -*-

;; Copyright (C) 20 Fermin Munoz

;; License: GPL-3.0-or-later

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

;;;; The requires
(require 'ert)

(provide 'test-islisp-mode.el)
;;; test-islisp-mode.el ends here


inferior-islisp.el
;;; inferior-islisp.el ---  Run inferior ISLisp processes            -*- lexical-binding: t; -*-

;; Copyright (C) 2021 Fermin Munoz

;; Author: Fermin Munoz
;; Maintainer: Fermin Munoz <fmfs@posteo.net>
;; Created: 24 Sep 2021
;; Version: 0.2.0
;; Keywords: islisp, lisp, programming
;; URL: https://gitlab.com/sasanidas/islisp-mode
;; Package-Requires: ((emacs "26.3"))
;; License: GPL-3.0-or-later

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Quick intro
;;

;;; Code:

;;;; The requires

(require 'comint)
(require 'islisp-mode)
(require 'inf-lisp)

(defgroup inferior-islisp nil
  "Run a ISLisp process in a buffer."
  :group 'islisp)

(defcustom inferior-islisp-command-line '("eisl")
  "Command line for calling an inferior islisp process."
  :type 'string
  :group 'inferior-islisp)

(defcustom inferior-islisp-source-modes '(islisp-mode)
  "List of modes which indicate a buffer contains ISLisp source code."
  :type '(repeat function)
  :group 'inferior-islisp)

(defcustom inferior-islisp-prompt "^>>? *"
  "Command line for calling an inferior islisp process."
  :type 'string
  :group 'inferior-islisp)

(defcustom inferior-islisp-load-command "(load \"%s\")\n"
  "Format-string for building a ISLisp expression to load a file."
  :type 'string
  :group 'inferior-islisp)

(defvar inferior-islisp-mode-hook '()
  "Hook for customising Inferior ISLisp mode.")

(defvar inferior-islisp-buffer nil "")

(defvar inferior-islisp-prev-l-c-dir-file nil
  "Record last directory and file used in loading or compiling.
This holds a cons cell of the form `(DIRECTORY . FILE)'
describing the last `islisp-load-file' or `islisp-compile-file' command.")

(defun inferior-islisp-proc ()
  (let ((proc (get-buffer-process (get-buffer inferior-islisp-buffer))))
    (or proc
	(error "No ISLisp subprocess; see variable `inferior-islisp-buffer'"))))

(defun inferior-islisp-eval-region (start end &optional _and-go)
  "Send the current region from START to END the inferior ISLisp process."
  (interactive "r\nP")
  (comint-send-region (inferior-islisp-proc) start end)
  (comint-send-string (inferior-islisp-proc) "\n"))

(defun inferior-islisp-eval-string (string)
  "Send STRING to the inferior Lisp process to be executed."
  (comint-send-string (inferior-islisp-proc) (concat string "\n")))

(defun inferior-islisp-do-defun (do-region)
  "Send the current defun in DO-REGION to the inferior Lisp process."
  (save-excursion
    (beginning-of-defun)
    (forward-sexp)
    (let ((end (point)) (case-fold-search t))
      (beginning-of-defun)
      (funcall do-region (point) end))))

(defun inferior-islisp-eval-defun (&optional _and-go)
  "Send the current defun to the inferior ISLisp process."
  (interactive "P")
  (islisp-do-defun 'inferior-islisp-eval-region))

(defun inferior-islisp-eval-last-sexp (&optional _and-go)
  "Send the previous sexp to the inferior ISLisp process."
  (interactive "p")
  (let* ((close (point-max))
	 (start (save-excursion
		  (when (and (eolp) (not (bolp)))
		    (backward-char))
		  (setq close (1- (scan-lists (point) 1 1)))
		  (when (< close (line-end-position))
		    (goto-char (1+ close))
		    (backward-list)
		    (point)))))
    (comint-send-string (inferior-islisp-proc)
			(buffer-substring-no-properties
			 start
			 (+ (point) 1)))
    (comint-send-string (inferior-islisp-proc) "\n")))

;;TODO: This is Easy-ISLisp specific
(defun inferior-islisp-load-file (file-name)
  "Load a ISLisp file with FILE-NAME into the inferior ISLisp process."
  (interactive "f")
  (comint-check-source file-name) 
  (setq inferior-islisp-prev-l-c-dir-file (cons (file-name-directory    file-name)
						(file-name-nondirectory file-name)))
  (comint-send-string (inferior-islisp-proc)
		      (format inferior-islisp-load-command file-name))

  (comint-send-string (inferior-islisp-proc)
		      (format "(format (standard-output) \"Loaded: %s\")\n" file-name)))

;;TODO: This is Easy-ISLisp specific
(defun inferior-islisp-compile-file (file-name)
  "Compile a ISLisp FILE-NAME in the inferior ISLisp process."
  (interactive (comint-get-source "Compile ISLisp file: " inferior-islisp-prev-l-c-dir-file
				  inferior-islisp-source-modes nil)) ; nil = don't need
					; suffix .lisp
  (comint-check-source file-name) ; Check to see if buffer needs saved.
  (setq inferior-islisp-prev-l-c-dir-file (cons (file-name-directory    file-name)
						(file-name-nondirectory file-name)))
  (comint-send-string (inferior-islisp-proc) (concat "(compile-file \"" file-name "\")\n")))

(defun inferior-islisp ()
  "Launch the inferior-islisp comint buffer."
  (interactive)
  (when (not (comint-check-proc "*inferior-islisp*"))
    (set-buffer (apply (function make-comint)
		       "inferior-islisp" inferior-islisp-command-line))
    (inferior-islisp-mode))
  (setq inferior-islisp-buffer "*inferior-islisp*")
  (pop-to-buffer-same-window "*inferior-islisp*"))

;; islisp-mode keybindings
(define-key islisp-mode-map (kbd "M-C-x") 'inferior-islisp-eval-defun)
(define-key islisp-mode-map (kbd "C-x C-e") 'inferior-islisp-eval-last-sexp)
(define-key islisp-mode-map (kbd "C-c C-r") 'inferior-islisp-eval-region)
(define-key islisp-mode-map (kbd "C-c C-l") 'inferior-islisp-load-file)
(define-key islisp-mode-map (kbd "C-c C-k") 'inferior-islisp-compile-file)

(define-derived-mode inferior-islisp-mode comint-mode "Inferior ISLisp"
  ""
  (islisp-mode-variables)
  (setq-local comint-prompt-regexp inferior-islisp-prompt
	      mode-line-process '(":%s")
	      comint-get-old-input (function lisp-get-old-input)
	      comint-prompt-read-only t)
  (islisp-use-implementation))


(provide 'inferior-islisp)
;;; inferior-islisp.el ends here


advance/islisp-tags.el
;;; islisp-tags.el ---  ISLisp tags support            -*- lexical-binding: t; -*-

;; Copyright (C) 2021 Fermin Munoz

;; Author: Fermin Munoz
;; Maintainer: Fermin Munoz <fmfs@posteo.net>
;; Created: 24 Sep 2021
;; Version: 0.2.0
;; Keywords: islisp, lisp, programming
;; URL: https://gitlab.com/sasanidas/islisp-mode
;; Package-Requires: ((emacs "26.3"))
;; License: GPL-3.0-or-later

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Quick intro
;;
(require 'etags)
(require 'project)
(require 'cl-lib)

;;; Code:

;; (defcustom islisp-tags-regex '("'{lisp}/[ \t]*defglobal_[A-Z_ \t(]+\"\([^\"]+\)\"/\1/'"
;; 			       "'{lisp}/[ \t]*defdynamic_[A-Z_ \t(]+\"\([^\"]+\)\"/\1/'")
;;   "ISLisp TAGS regex, they are use to create/update the TAGS."
;;   :type 'list
;;   :group 'islisp)

(defcustom islisp-tags-tag-library t
  "Generate tags for library directory."
  :type 'boolean
  :group 'islisp)

;; (defun islisp-tags--regex ()
;;   (let (regex-str)
;;     (cl-loop for reg in islisp-tags-regex
;; 	     do (setf regex-str
;; 		      (concat regex-str " " (format "--regex=%s" reg))))
;;     regex-str))

(defun islisp-tags--generate-in (directory &optional tags-location append)
  (let ((etags-format "etags --language=lisp %s" ))
    (when tags-location (setf etags-format (concat etags-format
						   (format " -o %s" tags-location))))
    (when append (setf etags-format (concat etags-format " --append")))
    (shell-command-to-string (format "cd %s && find . -name \"*.lsp\" -print | %s -"
				     directory
				     etags-format))))
(defun islisp-tags-generate ()
  "Generate current project tags, created with `etags'."
  (interactive)
  (let* ((project (project-current t)))
    (islisp-tags--generate-in (cdr project))
    (when islisp-tags-tag-library
      (islisp-tags--update-library (cdr project)))))

(defun islisp-tags--beginning ()
  "Return the name of the position."
  (save-excursion
    (backward-word)
    (point)))

(defun islisp-tags-autocomplete ()
  "Show auto-completion at point using current tags table."
  (interactive)
  (let ((current-prefix (buffer-substring-no-properties (islisp-tags--beginning) (point)))
	(tags-table (tags-completion-table)))
    (completion-in-region (islisp-tags--beginning) (point) tags-table)))

(defun islisp-tags-symbols-navigate ()
  "Find the workspace TAG definition."
  (interactive)
  (xref-find-definitions
   (completing-read "Select tag symbol:" (tags-completion-table))))


(defun islisp-tags--update-library (root)
  (islisp-tags--generate-in islisp-library-directory (concat root "TAGS") t))

(define-key islisp-mode-map (kbd "C-c C-w") 'islisp-tags-symbols-navigate)
(define-key islisp-mode-map (kbd "C-c TAB") 'islisp-tags-autocomplete)
(define-key islisp-mode-map (kbd "C-c C-e") 'islisp-tags-generate)




;;;; The requires
(provide 'islisp-tags)
;;; islisp-tags.el ends here


islisp-mode.el
;;; islisp-mode.el ---  Major mode for ISLisp programming            -*- lexical-binding: t; -*-

;; Copyright (C) 2021 Fermin Munoz

;; Author: Fermin Munoz
;; Maintainer: Fermin Munoz <fmfs@posteo.net>
;; Created: 24 Sep 2021
;; Version: 0.2.0
;; Keywords: islisp, lisp, programming
;; URL: https://gitlab.com/sasanidas/islisp-mode
;; Package-Requires: ((emacs "26.3"))
;; License: GPL-3.0-or-later

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Quick intro
;;

;;; Code:

;;;; The requires

(require 'font-lock)
(require 'lisp-mode)
(require 'islisp-hyperdraft)

(defgroup islisp nil
  "ISLisp support."
  :group 'languages)

;;TODO: Make this a list.
(defcustom islisp-current-implementation 'easy-islisp
  "Current mode implementation.
Each implementation may add custom functionality,
the supported implementations are define in the implementations
directory."
  :type 'symbol
  :group 'islisp)

(defcustom islisp-advance-features nil
  "Enable advance IDE-like features like autocompletion."
  :type 'symbol
  :group 'islisp)

(defun islisp-use-implementation ()
  "Initialise the `islisp-current-implementation'."
  (funcall #'require islisp-current-implementation)
  (funcall (intern
	    (apply 'concat `(,(format "%S" islisp-current-implementation) "-init")))))

(defvar islisp-mode-map (make-sparse-keymap))

;;(define-key islisp-mode-map (kbd "C-c C-d") 'islisp-describe-symbol)

(define-key islisp-mode-map (kbd "C-c C-i") 'islisp-repl)
(define-key islisp-mode-map (kbd "C-c M-q") 'indent-region)
(define-key islisp-mode-map (kbd "C-c C-d") 'islisp-hyperdraft-lookup-documentation)


(defun islisp--create-mode-menu ()
  "Internal function to create or recreate the plisp-mode menu."
  (easy-menu-define islisp-menu islisp-mode-map "Menu bar entry for `islisp-mode'"
    '("ISLisp"
      ["Eval last sexp" islisp-eval-last-sexp :keys "C-x C-e"]
      ["Eval function" islisp-eval-defun t  :keys "M-C-x"]
      ["Eval region" islisp-eval-region t  :keys "M-C-x"]
      ["Lookup Documentation" islisp-lookup-documentation t  :keys "C-c C-d"]
      "--"
      ["ISLisp REPL" islisp-repl t :keys "C-c C-i"]
      ["Comment/Uncomment region" comment-line t :keys "C-x C-;"]
      ["Load file" islisp-load-file t  :keys "C-c C-l"]
      ["Compile file" islisp-compile-file t :keys "C-c C-k"]
      "--")))

(defconst islisp-mode-symbol-regexp lisp-mode-symbol-regexp)

(defun islisp-mode-variables ()
  "ISLisp major mode default variables."
  (set-syntax-table lisp-mode-syntax-table)
  (setq-local paragraph-ignore-fill-prefix t
	      fill-paragraph-function 'lisp-fill-paragraph
	      adaptive-fill-function #'lisp-adaptive-fill
	      indent-line-function 'lisp-indent-line
	      indent-region-function 'lisp-indent-region
	      comment-indent-function #'lisp-comment-indent
	      outline-level 'lisp-outline-level
	      add-log-current-defun-function #'lisp-current-defun-name
	      comment-start ";"
	      comment-start-skip ";+ *"
	      comment-add 1
	      comment-column 40
	      comment-use-syntax t
	      multibyte-syntax-as-symbol t
	      completion-ignored-extensions (remove ".o" completion-ignored-extensions)
	      imenu-generic-expression islisp-imenu-generic-expression)
  (islisp-set-fl-keys)
  (islisp--create-mode-menu))

(defvar islisp-imenu-generic-expression
  (list
   (list (purecopy "Functions")
	 (purecopy (concat "^\\s-*("
			   (regexp-opt
			    '("defun")
			    t)
			   "\\s-+\\(" islisp-mode-symbol-regexp "\\)"))
	 2)
   (list (purecopy "Generics/Methods")
	 (purecopy (concat "^\\s-*("
			   (regexp-opt
			    '("defgeneric" "defmethod")
			    t)
			   "\\s-+\\(" islisp-mode-symbol-regexp "\\)"))
	 2)
   (list (purecopy "Macros")
	 (purecopy (concat "^\\s-*("
			   (regexp-opt
			    '("defmacro")
			    t)
			   "\\s-+\\(" islisp-mode-symbol-regexp "\\)"))
	 2)
   (list (purecopy "Variables")
	 (purecopy (concat "^\\s-*("
			   (eval-when-compile
			     (regexp-opt
			      '("defglobal" "")
                              t))
			   "\\s-+\\(" islisp-mode-symbol-regexp "\\)"))
	 2)
   (list (purecopy "Types")
	 (purecopy (concat "^\\s-*("
			   (eval-when-compile
			     (regexp-opt
			      '("defclass")
                              t))
			   "\\s-+'?\\(" islisp-mode-symbol-regexp "\\)"))
	 2))
  "Imenu generic expression for ISLisp mode.  See `imenu-generic-expression'.")


(defun islisp-set-fl-keys ()
  (setq-local
   font-lock-defaults `(islisp-font-lock-keywords
		nil nil nil nil
		(font-lock-syntactic-face-function
		 . lisp-font-lock-syntactic-face-function))))

(defvar islisp-general-keywords
  '("-" "*" "/=" "+" "<" "<=" "=" ">" ">="
    "abs" "append" "apply" "aref" "arithmetic-error-operands"
    "arithmetic-error-operation" "array-dimensions" "assoc" "atan"
    "atan2" "atanh" "atom" "basic-array-p" "basic-array*-p"
    "basic-vector-p" "call-next-method" "car" "cdr" "ceiling"
    "char-index" "char/=" "char<" "char<=" "char="
    "char>" "char>=" "characterp" "class-of" "close"
    "condition-continuable" "cons" "consp" "continue-condition"
    "cos" "cosh" "create-array" "create-list"
    "create-string-input-stream"
    "create-string-output-stream" "create-string" "create-vector"
    "create"
    "div" "domain-error-object" "domain-error-expected-class"
    "dummyp" "elt" "eq" "eql" "equal" "error-output" "error"
    "eval" "exp" "expt" "file-length" "file-position" "finish-output"
    "float" "floatp" "floor" "format-char" "format-fresh-line"
    "format-float" "format-integer" "format-object" "format-tab"
    "format"
    "funcall" "functionp" "garef" "gbc" "gcd" "general-array*-p"
    "general-vector-p" "generic-function-p" "gensym"
    "get-internal-real-time"
    "get-internal-run-time"
    "get-output-stream-string" "get-universal-time" "hdmp" "identity"
    "initialize-object" "input-stream-p" "instancep" "integerp"
    "internal-time-units-per-second" "isqrt" "lcm" "length" "list"
    "listp" "load" "log" "map-into" "mapc" "mapcar" "mapcan"
    "mapcon" "mapl" "maplist" "max" "member" "min" "mod"
    "next-method-p" "not" "nreverse" "null" "numberp"
    "open-input-file" "open-io-file" "open-output-file" "open-stream-p"
    "output-stream-p" "parse-error-string" "parse-error-expected-class"
    "parse-number" "preview-char" "prin1" "print" "probe-file"
    "property" "quit" "quotient" "read-byte" "read-char" "read-line"
    "read" "reciprocal" "remove-property" "reverse" "round"
    "set-aref"
    "set-car" "set-cdr" "set-elt" "set-file-position" "set-garef"
    "set-property" "signal-condition" "simple-error-format-argument"
    "simple-error-format-string" "sin" "sinh" "slot-value" "sqrt"
    "standard-input" "standard-output" "stream-error-stream" "streamp"
    "stream-ready-p" "string-append" "string-index" "string/="
    "string<" "string<=" "string=" "string>" "string>=" "stringp"
    "subclassp"
    "subseq" "symbolp" "tan" "tanh" "truncate"
    "undefined-entity-name"
    "undefined-entity-namespace" "vector" "write-byte" "import"
    "lambda" "labels" "flet" "let" "let*" "setq" "setf"
    "dynamic" "set-dynamic" "function" "function*" "symbol-function" "class"
    "and" "or" "if" "cond" "while" "for" "block" "return-from"
    "case" "case-using" "progn" "dynamic-let" "ignore-errors" "catch" "throw"
    "tagbody" "go" "unwind-protect" "with-standard-input"
    "with-standard-output" "with-error-output" "with-handler"
    "convert" "with-open-input-file" "with-open-output-file"
    "with-open-io-file" "the" "assure"))

(defvar islisp-font-lock-keywords-regex
  (regexp-opt islisp-general-keywords t))

(defvar islisp-def-keywords
  '("defconstant" "defglobal" "defdynamic"))

(defvar islisp-def-keywords-regex
  (regexp-opt islisp-def-keywords t))

(defvar islisp-func-def-keywords
  '("defun" "defmethod" "defmacro" "defgeneric" "defclass" "defgeneric*"))

(defvar islisp-func-def-keywords-regex
  (regexp-opt islisp-func-def-keywords t))

(defvar islisp-font-lock-keywords
  `((,(concat "(" islisp-font-lock-keywords-regex "\\_>")
     (1 'font-lock-keyword-face t))
    (,(concat "\\_<:" islisp-mode-symbol-regexp "\\_>")
     (0 'font-lock-type-face t))
    (,(concat "(" (regexp-opt '("cerror") t) "\\_>")
     (1 'font-lock-warning-face))
    (,(concat "(" islisp-def-keywords-regex "\\_>"
	      ;; Any whitespace and defined object.
	      "[ \t']*"
	      "\\(([ \t']*\\)?" ;; An opening paren.
	      "\\(\\(setf\\)[ \t]+" lisp-mode-symbol-regexp
	      "\\|" lisp-mode-symbol-regexp "\\)?")
     (1 'font-lock-keyword-face)
     (3 'font-lock-variable-name-face nil t))
    (,(concat "(" islisp-func-def-keywords-regex "\\_>"
	      ;; Any whitespace and defined object.
	      "[ \t']*"
	      "\\(([ \t']*\\)?" ;; An opening paren.
	      "\\(\\(setf\\)[ \t]+" lisp-mode-symbol-regexp
	      "\\|" lisp-mode-symbol-regexp "\\)?")
     (1 'font-lock-keyword-face)
     (3 'font-lock-function-name-face nil t))))

(defun islisp-repl()
  "Start a ISLisp REPL or switch to it."
  (interactive)
  (require 'inferior-islisp)
  (if (and (stringp inferior-islisp-buffer)
	   (get-buffer inferior-islisp-buffer))
      (pop-to-buffer inferior-islisp-buffer)
    (inferior-islisp)))

;;;###autoload
(define-derived-mode islisp-mode prog-mode "ISLisp"
  "Major mode for editing ISLisp code"
  (islisp-mode-variables)
  (setq-local
   find-tag-default-function 'lisp-find-tag-default
   comment-start-skip
   "\\(\\(^\\|[^\\\\\n]\\)\\(\\\\\\\\\\)*\\)\\(;+\\|#|\\) *")
  (when islisp-advance-features
    (require 'islisp-tags)
    (easy-menu-add-item 'islisp-menu nil
			'("Tags"
			  ["Generate TAGS"  islisp-tags-generate t :keys "C-c g"]
			  ["Symbols navigation" islisp-tags-symbols-navigate t  :keys "C-c w"]
			  ["Tag auto-complete" islisp-tags-autocomplete t  :keys "C-c TAB"])))
  (islisp-use-implementation))

(advice-add 'islisp-mode :after #'(lambda ()
			      (setq-local font-lock-keywords-case-fold-search t)))

(add-to-list 'auto-mode-alist '("\\.lsp\\'" . islisp-mode))


(provide 'islisp-mode)
;;; islisp-mode.el ends here
